examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+		auto vk_surface = core::smart_refctd_ptr_static_cast<video::ISurfaceVK>(surface);
+
+		bool isGPUSuitable = false;
+
+		// Check if the physical device has queue families which support both graphics and presenting, not necessarily overlapping
+		{
+			const auto& queueFamilyProperties = gpu->getQueueFamilyProperties();
+
+			for (uint32_t familyIndex = 0u; familyIndex < queueFamilyProperties.size(); ++familyIndex)
+			{
+				const auto& familyProperty = queueFamilyProperties.begin() + familyIndex;
+				if (familyProperty->queueFlags & video::IPhysicalDevice::E_QUEUE_FLAGS::EQF_GRAPHICS_BIT)
+					graphicsFamilyIndex = familyIndex;
+
+				if (surface->isSupported(gpu.get(), familyIndex))
+					presentFamilyIndex = familyIndex;
+
+				if ((graphicsFamilyIndex != ~0u) && (presentFamilyIndex != ~0u))
+				{
+					isGPUSuitable = true;
+					break;
+				}
+			}
+		}
+
+		// Check if this physical device supports the swapchain extension
+		// Todo(achal): Eventually move this to somewhere inside the engine
+		{
+			// Todo(achal): Get this from the user
+			const uint32_t requiredDeviceExtensionCount = 1u;
+			const char* requiredDeviceExtensionNames[] = { VK_KHR_SWAPCHAIN_EXTENSION_NAME };
 
+			uint32_t availableExtensionCount;
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, NULL);
+			std::vector<VkExtensionProperties> availableExtensions(availableExtensionCount);
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, availableExtensions.data());
+
+			bool requiredDeviceExtensionsAvailable = false;
+			for (uint32_t i = 0u; i < availableExtensionCount; ++i)
+			{
+				if (strcmp(availableExtensions[i].extensionName, requiredDeviceExtensionNames[0]) == 0)
+				{
+					requiredDeviceExtensionsAvailable = true;
+					break;
+				}
+			}
+
+			if (!requiredDeviceExtensionsAvailable)
+				isGPUSuitable = false;
+		}
+
+		// Check if the surface is adequate
+		{
+			std::vector<video::ISurface::SFormat> surfaceFormats =
+				gpu->getAvailableFormatsForSurface(surface.get());
+
+			std::vector<video::ISurface::E_PRESENT_MODE> presentModes =
+				gpu->getAvailablePresentModesForSurface(surface.get());
+
+			// Todo(achal): Probably should make a ISurface::SCapabilities
+			// struct for this as a wrapper for VkSurfaceCapabilitiesKHR
+			// nbl::video::ISurface::SCapabilities surfaceCapabilities = ;
+			VkSurfaceCapabilitiesKHR surfaceCapabilities;
+			vkGetPhysicalDeviceSurfaceCapabilitiesKHR(gpu->getInternalObject(),
+				vk_surface->m_surface, &surfaceCapabilities);
+
+			if ((surfaceCapabilities.maxImageCount != 0) && (SC_IMG_COUNT > surfaceCapabilities.maxImageCount)
+				|| (surfaceFormats.empty()) || (presentModes.empty()))
+			{
+				isGPUSuitable = false;
+			}
+
+			// Todo(achal): Probably a more sophisticated way to choose these
+			surfaceFormat = surfaceFormats[0];
+			presentMode = presentModes[0];

```

Comment:
I'd recommend some CRTP in CommonAPI and make static (or not) methods for choosing surfaceFormat and presentMode

(default choice should be RGB8_SRGB, then failing RGBA8_SRGB, then failing RGB10A2, then failing RGBA16_SFLOAT)
(default presentMode should be FIFO_RELAXED if supported, otherwise fallback to no Vsync)

Reply:
P.S. Dont forget to query and report the colorspace and OETF.

Reply:
I think @Erfan-Ahmadi wrote this part of the CommonAPI

Reply:
@devshgraphicsprogramming We now get the user requested format and colorspace and there is a kind of temporary fix for Vulkan which tries to find the nearest match by finding the matching format -> color_space -> wider_bit_depth (in that order)
I think It's better If I defer this to later; to work on it alongside my format promotion work, since I'll be doing similar things and will be more in the mindset

Reply:
fixed by @deprilula28